Add support for llvm.coverage.json.export format

paulofaria opened the pull request:
This PR adds support for the llvm.coverage.json.export format. This is the format generated by the Swift Package Manager. The commands below allow you to generate the coverage file:

```console
mkdir TestCoverage
cd TestCoverage
swift package init
swift test --enable-code-coverage  
swift test --show-codecov-path
```

The last command shows the path of the generated coverage file. I added a new formatter called `lcov-json` which can be used like so:

```console
test-reporter format-coverage $(swift test --show-codecov-path) -t lcov-json    
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/codeclimate/test-reporter?pullRequest=439) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/codeclimate/test-reporter?pullRequest=439) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/codeclimate/test-reporter?pullRequest=439) it.</sub>

paulofaria commented:
@filipesperandio Hey, Filipe. Hope everything is well with you and your loved ones. Can you please take a look at the code and let me know if there is anything I need to rework? Thank you!

fede-moya commented:
@filipesperandio It works as described by @paulofaria, attaching screenshot. 
<img width="1158" alt="Screen Shot 2020-09-07 at 19 16 40" src="https://user-images.githubusercontent.com/11605222/92418604-c1358d00-f13e-11ea-8892-32304c8d6cd6.png">

fede-moya commented:
could you @paulofaria got it to work using the `after-build` command ?

filipesperandio commented:
@fede-moya Is there an error with the `after-build`?

fede-moya commented:
yeap, there is an error.

```sh
swift-llvm-cov-test >> test-reporter after-build -t lcov-json
ERRO[0000] could not find coverage file
could not find any files in search paths for lcov-json. search paths were:
Error: could not find any files in search paths for lcov-json. search paths were:
Usage:
  cc-test-reporter after-build [flags]

Flags:
  -s, --batch-size int               batch size for source files (default 500)
  -e, --coverage-endpoint string     endpoint to upload coverage information to (default "https://api.codeclimate.com/v1/test_reports")
  -t, --coverage-input-type string   type of input source to use [clover, cobertura, coverage.py, excoveralls, gcov, gocov, jacoco, lcov, lcov-json, simplecov, xccov]
      --exit-code int                exit code of the test run
  -r, --id string                    reporter identifier (default "41241242141")
      --insecure                     send coverage insecurely (without HTTPS)
  -p, --prefix string                the root directory where the coverage analysis was performed (default "/Users/federicomoya/Documents/repos/swift-llvm-cov-test")

Global Flags:
  -d, --debug   run in debug mode
```

```
swift-llvm-cov-test >> ls
Package.swift README.md     Sources       Tests         coverage
swift-llvm-cov-test >> ls coverage
swift-llvm-cov-test.json
```

filipesperandio commented:
Thanks @fede-moya .
@paulofaria Any idea?

paulofaria commented:
There is no default search paths for coverage files generated by the Swift Package Manager. As you can see I removed the default search path in this [commit](https://github.com/codeclimate/test-reporter/pull/439/commits/8fad4e15bf8708d790a22583071b1a9d4a001c1a) (the one I removed was a leftover from the formatter I used as a base for this formatter).  The Swift Package Manager code coverage feature was designed to be used with `swift test --show-codecov-path`, so I don't think it's possible to use it with the `after-build` command.

paulofaria commented:
Just inspected the directory of the code coverage file created by SwiftPM. Running the following command:

```
swift test --show-codecov-path
```

Gives us this output in my machine:

```
/Users/paulo/Development/Playgrounds/TestCoverage/.build/x86_64-apple-macosx/debug/codecov/TestCoverage.json
```

Inspecting the directory of the file with: 

```
ls /Users/paulo/Development/Playgrounds/TestCoverage/.build/x86_64-apple-macosx/debug/codecov
```

We get:

```
TestCoverage.json                    default1567809077063180439_0.profraw
default.profdata
```

Does default search path support globbing? If so, we could try to match any `.json` file inside a `codecov` directory.

paulofaria commented:
I should warn that, since this approach assumes implementation details of SwiftPM, it is not future-proof. SwiftPM might change the directory and filename of the code coverage file.

paulofaria commented:
Looks like the CI error is unrelated to the changes in this PR.

```
/usr/local/Homebrew/Library/Homebrew/brew.rb:16:in `<main>': HOMEBREW_REQUIRED_RUBY_VERSION was not exported! Please call bin/brew directly! (RuntimeError)
```

fede-moya commented:
@filipesperandio @paulofaria [this commit](https://github.com/codeclimate/test-reporter/pull/441/commits/7ab1e61ad5c4bcff6f04176870f36e593f15bb01) fixes the CI. It should get merge soon. 

Considering that according to @paulofaria due to the specifics of swift the `after build` command doesn't make much sense, having just the option to use this reporter with the `before build` command sounds good to me. Although we should document that. 

I think after the fix of the CI gets in, we could merge this one. WDYT @filipesperandio  ?

noelia-lencina commented:
@fede-moya @filipesperandio pinging you as you have been following this PR closer. The CI update did indeed fix the failing CI in this PR, do you think this is safe to merge or more steps are needed?

fede-moya commented:
> @fede-moya @filipesperandio pinging you as you have been following this PR closer. The CI update did indeed fix the failing CI in this PR, do you think this is safe to merge or more steps are needed?

Nice, I'm good with how it is now. The only thing would be for us to add a ticket to mention this new option in the docs, and it's particular use, like mentioning that there is no expected path.
